src/io/mod.rs

```diff
@@ -0,0 +1,202 @@
+        }
+
+        fn build(self) -> LocalFsReader {
+            LocalFsReader::new(self)
+        }
+    }
+
+    impl ReaderConfiguration for LocalFsReaderConfig {}
+
+    /// Reads all files specified in a given directory from the local directory
+    /// on all executors on every worker node.
+    struct LocalFsReader {
+        path: PathBuf,
+        is_single_file: bool,
+        filter_ext: Option<std::ffi::OsString>,
+        expect_dir: bool,
+        files: Vec<Vec<PathBuf>>,
+        open_file: Option<BufReader<fs::File>>,
+    }
+
+    // TODO: optimally read should be async so the executors can do other work meanwhile,
+    // and shouldn't run too many parallel reads per worker node as it's wasteful.
+    impl LocalFsReader {
+        pub fn new(config: LocalFsReaderConfig) -> LocalFsReader {
+            let LocalFsReaderConfig {
+                dir_path,
+                expect_dir,
+                filter_ext,
+            } = config;
+
+            let is_single_file = {
+                let path: &Path = dir_path.as_ref();
+                path.is_file()
+            };
+
+            LocalFsReader {
+                path: dir_path,
+                is_single_file,
+                filter_ext,
+                expect_dir,
+                files: vec![],
+                open_file: None,
+            }
+        }
+
+        fn load_files(&mut self) -> Result<()> {
+            let mut total_size = 0_u64;
+            if self.is_single_file {
+                let size = fs::metadata(&self.path)?.len();
+                total_size += size;
+                self.files.push(vec![self.path.clone()]);
+                return Ok(());
+            }
+
+            let num_partitions = 0;
+
+            let mut files: Vec<(u64, PathBuf)> = vec![];
+            // We compute std deviation incrementally to estimate a good breakpoint
+            // of size per partition.
+            let mut total_files = 0_u64;
+            let mut k = 0;
+            let mut ex = 0.0;
+            let mut ex2 = 0.0;
+
+            for (i, entry) in fs::read_dir(&self.path)?.enumerate() {
+                let path = entry?.path();
+                if path.is_file() {
+                    if !self.is_proper_file(&path) {
+                        continue;
+                    }
+                    let size = fs::metadata(&path)?.len();
+                    if i == 0 {
+                        // assign first file size as reference sample
+                        k = size;
+                    }
+                    // compute the necessary statistics
+                    ex += (size - k) as f32;
+                    ex2 += (size - k) as f32 * (size - k) as f32;
+                    total_size += size;
+                    total_files += 1;
+
+                    files.push((size, path));
+                }
+            }
+
+            // Partition files according to total avg partition size and std dev of file size
```

Comment:
This function is very long by this point - it might be worth splitting the rest out from here so it's easily unit testable without any filesystem interaction.

Reply:
You checking an old commit, is split in the most recent one.